03.Isolation-by-environment_GDM
================

GDM for all P. turqueti samples:

``` bash
## Calculate pairwise genetic distance with ngsDist 
##From the step of "Filter sites with maximum observed heterozygosity of 0.5" in 01.SNP_filtering.Rmd
cd ./turqueti/tarcap/SNPfiltering1/9_max_obs_het05

/sw/containers/vcftools-0.1.16.sif vcftools --vcf ./turqueti/tarcap/SNPfiltering1/8_maf005/maf005_setID.vcf --exclude site_obs_het_over_05.tsv --out max_het05.vcf --recode --recode-INFO-all

## For all P. turqueti samples (don't forget to remove outgroup first)
module load htslib

bgzip max_het05.vcf
tabix max_het05.vcf.gz

bcftools view -f PASS --types snps -m1 -M2 -r `seq -s, 1 22` max_het05.vcf.gz | singularity run $SING/plink-1.90b6.26.sif plink1.9 --double-id --vcf max_het05.vcf.gz --recode A-transpose --allow-extra-chr --out ./ngsDist/allsamples/allsamples

tail -n +2 allsamples.traw | cut -f 1,4- | perl -p -e 's/\bNA\b/-1/g' | gzip > ngsDist.geno.gz

# Labels file
zgrep CHROM max_het05.vcf.gz | cut -f 10- | tr "\t" "\n" > ngsDist.labels

# Calculate pairwise distance with ngsDist 
ngsDist --geno ngsDist.geno.gz --n_ind 96 --n_sites 37322 --labels ngsDist.labels --out tar.dist 
```

``` r
#Import pairwise distance from ngsDist 
dist <- as.matrix(read.table("tar.dist"))

#Remove column 1 (arbitrary ID names generated from ngsDist)
dist <- subset(dist, select=-c(V1))

#Need to remove 3 samples as we only have 93 samples suitable for GDM for the analysis of all P. turqueti
#JS_40 - line 21
#JS_44 - line 22
#NIWA87970 - line 34

#remove column 21+1, 22+1, 34+1
dist_sub1 <- subset(dist, select=-c(V22, V23, V35))
#remove row 21, 22, 34
dist_93 <- dist_sub1[-c(21, 22, 34),]

#read in environmental variable data
tur_env_var_GDM <- read.csv("env_var_GDM.csv")

#grab ID column
tur_GDM_ID <- as.data.frame(tur_env_var_GDM$ID)
colnames(tur_GDM_ID) <- c("ID")

#run GDM
#code from https://github.com/pgugger/LandscapeGenomics/blob/master/2017/Exercise4.md#generalized-dissimilarity-modeling-gdm

gdm_cov <- cbind(tur_GDM_ID, dist_93)

#check min max value in gdm_cov
min(dist_93)
max(dist_93)

library(gdm)
gdm.input <- formatsitepair(bioData=gdm_cov, bioFormat=3, predData=tur_env_var_GDM, siteColumn="ID", XColumn="long", YColumn="lat")

gdm <- gdm(gdm.input, geo = T, splines = NULL, knots = NULL)
summary(gdm)

gdm$explained #29.90663

dev.new()
plot(gdm, plot.layout=c(3,3))
dev.off()

gdm.importance <- gdm.varImp(gdm.input, geo=T, nPerm=999)
dev.new()
barplot(sort(gdm.importance[[2]][,1], decreasing=T))
dev.off()
```

For P. turqueti samples only from the Scotia Sea

``` bash
## Calculate pairwise genetic distance with ngsDist 
##From the step of "Filter sites with maximum observed heterozygosity of 0.5" in 01.SNP_filtering.Rmd
/sw/containers/vcftools-0.1.16.sif vcftools --vcf max_het05.vcf --keep scotia_only.tsv --recode --recode-INFO-all --out max_het05_scotia

module load htslib

bgzip max_het05_scotia.vcf
tabix max_het05_scotia.vcf.gz

bcftools view -f PASS --types snps -m1 -M2 -r `seq -s, 1 22` max_het05_scotia.vcf.gz | singularity run $SING/plink-1.90b6.26.sif plink1.9 --double-id --vcf max_het05_scotia.vcf.gz --recode A-transpose --allow-extra-chr --out ./ngsDist/scotia/scotia_only

tail -n +2 scotia_only.traw | cut -f 1,4- | perl -p -e 's/\bNA\b/-1/g' | gzip > ngsDist_scotia.geno.gz

# Labels file
zgrep CHROM max_het05_scotia.vcf.gz | cut -f 10- | tr "\t" "\n" > ngsDist_scotia.labels

# Calculate pairwise distance with ngsDist 

ngsDist --geno ngsDist_scotia.geno.gz --n_ind 52 --n_sites 37322 --labels ngsDist_scotia.labels --out ./ngsDist/scotia/scotia_only/scotia.dist 
```

``` r
#Import pairwise distance from ngsDist 
dist <- as.matrix(read.table("scotia.dist"))

#read in environmental variable data
tur_env_var_GDM <- read.csv("scotia_env_var_GDM.csv")

#grab ID column
tur_GDM_ID <- as.data.frame(tur_env_var_GDM$ID)
colnames(tur_GDM_ID) <- c("ID")

#run GDM
#code from https://github.com/pgugger/LandscapeGenomics/blob/master/2017/Exercise4.md#generalized-dissimilarity-modeling-gdm

gdm_cov <- cbind(tur_GDM_ID, dist)

#check min max value in gdm_cov
min(dist)
max(dist)

library(gdm)
gdm.input <- formatsitepair(bioData=gdm_cov, bioFormat=3, predData=tur_env_var_GDM, siteColumn="ID", XColumn="long", YColumn="lat")

gdm <- gdm(gdm.input, geo = T, splines = NULL, knots = NULL)
summary(gdm)

gdm$explained #37.49608

dev.new()
plot(gdm, plot.layout=c(3,3))
dev.off()

gdm.importance <- gdm.varImp(gdm.input, geo=T, nPerm=999)
dev.new()
barplot(sort(gdm.importance[[2]][,1], decreasing=T))
dev.off()
```
