---
title: "04_RDA"
output: github_document
---

RDA for all P. turqueti samples:
```{R, error=TRUE, results='hide', eval=FALSE}
#RDA code from https://popgen.nescent.org/2018-03-27_RDA_GEA.html

library(psych)
library(vegan)
library(vcfR)
library(adegenet)
library(dartR)

#genotype file from -012 output of vcftools, using the thin1000_maf005_obshet05.vcf file, from 01.SNP_filtering.Rmd
geno_tur_tar_93 <- read.csv("./tur_tar_93_5188geno.csv", header=TRUE, row.names=1) 

str(geno_tur_tar_93)
dim(geno_tur_tar_93)

#RDA requires complete data frames (i.e., no missing data). Here to impute using the most common genotype at each SNP across all individuals

sum(is.na(geno_tur_tar_93)) #1963 NAs

geno_tur_tar_93.imp <- apply(geno_tur_tar_93, 2, function(x) replace(x, is.na(x), as.numeric(names(which.max(table(x))))))

sum(is.na(geno_tur_tar_93.imp)) # No NAs

#Read in the environmental factors
env <- read.csv("./tur_tar_env_var_GDM.csv")
str(env) # Look at the structure of the data frame
env$ID <- as.character(env$ID) # Make individual names characters (not factors)

#Confirm genotypes and environmental data are in the same order
identical(rownames(geno_tur_tar_93.imp), env[,2])
as.list(rownames(geno_tur_tar_93.imp))
as.list.data.frame(env[,2])
dim(env)

#check predictor correlations
png("./env_cor_panels.png", width=50, height=50, units="cm",res=155,pointsize = 26)
pairs.panels(env[,6:10], scale=T) #keep depth, longitude, bottom water salinity, bottom water temp_bottom, summer surface nitrate 
dev.off()

pred <- env[,6:10] #keep depth, longitude, bottom water salinity, bottom water temp_bottom, summer surface nitrate 

# Run the RDA

tur.rda <- rda(geno_tur_tar_93.imp ~ ., data=pred, scale=T)
tur.rda

RsquareAdj(tur.rda)

summary(eigenvals(tur.rda, model = "constrained"))
screeplot(tur.rda) 

#Check RDA significance 
signif.full <- anova.cca(tur.rda, parallel=getOption("mc.cores")) # default is permutation=999
signif.full 

signif.axis <- anova.cca(tur.rda, by="axis", parallel=getOption("mc.cores"))
signif.axis

#Double check multicollinearity between predictors again 
vif.cca(tur.rda)

#Plot the RDA
plot(tur.rda, scaling=3) 

#Plot RDA by shelf/scotia
levels(env$region) <- c("shelf", "scotia")
env$region <- as.factor(env$region)
spe <- env$region
bg <- c('#1b9e77', '#d95f02')

#axis 1 &2 
plot(tur.rda, type="n", scaling=3, family="sans", cex=0.75) 
points(tur.rda, display="species", pch=20, cex=0.7, col="gray32", scaling=3)  # the SNPs 
points(tur.rda, display="sites", pch=21, cex=1.3, col="gray32", scaling=3, bg=bg[spe])  # shelf & scotia +
text(tur.rda, scaling=3, display="bp", col="black", cex=0.7, family="sans")                           # the predictors
legend("bottomright", legend=levels(spe), bty="n", col="gray32", pch=21, cex=0.75, pt.bg=bg)

# axes 3 & 4
plot(tur.rda, type="n", scaling=3, choices=c(3,4))
points(tur.rda, display="species", pch=20, cex=0.7, col="gray32", scaling=3, choices=c(3,4))
points(tur.rda, display="sites", pch=21, cex=1.3, col="gray32", scaling=3, bg=bg[spe], choices=c(3,4))
text(tur.rda, scaling=3, display="bp", col="black", cex=0.75, family="sans", choices=c(3,4))
legend("bottomright", legend=levels(spe), bty="n", col="gray32", pch=21, cex=0.75, pt.bg=bg)

#Plot RDA by locations
env$location <- factor(env$location,levels=c("ShagRocks","SouthGeorgia","SouthOrkney","ElephantIs","BransfieldStrait","KingGeorgeIs","RobertIs","LivingstonIs","DeceptionIs","Peninsula", "SouthWeddellSea", "EastWeddellSea","Amundsen","RossSea","Adelie","Casey","Prydz"))
spe <- env$location
bg <- c('#e6194B', '#3cb44b', '#ffe119', '#4363d8', '#f58231', '#42d4f4', '#f032e6', '#fabed4', '#469990', '#dcbeff', '#9A6324', '#fffac8', '#800000', '#aaffc3', '#000075', '#a9a9a9', '#ffffff', '#000000')

#axis 1 &2 
plot(tur.rda, type="n", scaling=3, family="sans", cex=0.75) 
points(tur.rda, display="species", pch=20, cex=0.7, col="gray32", scaling=3)  # the SNPs 
points(tur.rda, display="sites", pch=21, cex=1.3, col="gray32", scaling=3, bg=bg[spe])  # shelf & scotia +
text(tur.rda, scaling=3, display="bp", col="black", cex=0.7, family="sans")                           # the predictors
legend("bottomright",  legend=levels(spe), xpd=TRUE, bty="n", col="gray32", pch=21, cex=0.75, pt.bg=bg)

#Identify candidate outlier SNPs

load.rda <- scores(tur.rda, choices=c(1:4), display="species")  # Species scores for the first constrained axes

#Axis 1-4 were selected as they were signficiant 

hist(load.rda[,1], main="Loadings on RDA1")
hist(load.rda[,2], main="Loadings on RDA2")
hist(load.rda[,3], main="Loadings on RDA3")
hist(load.rda[,4], main="Loadings on RDA4")

#define the function here as outliers

outliers <- function(x,z){
  lims <- mean(x) + c(-1, 1) * z * sd(x)     # find loadings +/-z sd from mean loading     
  x[x < lims[1] | x > lims[2]]               # locus names in these tails
}

#Apply outlier func to each significant constrained axis (3 SD)
cand1 <- outliers(load.rda[,1],3) # 29
cand2 <- outliers(load.rda[,2],3) # 55
cand3 <- outliers(load.rda[,3],3) # 10
cand4 <- outliers(load.rda[,4],3) # 13

ncand <- length(cand1) + length(cand2) + length(cand3) + length(cand4) 
ncand #110 candidate outliers

cand1 <- cbind.data.frame(rep(1,times=length(cand1)), names(cand1), unname(cand1))
cand2 <- cbind.data.frame(rep(2,times=length(cand2)), names(cand2), unname(cand2))
cand3 <- cbind.data.frame(rep(3,times=length(cand3)), names(cand3), unname(cand3))
cand4 <- cbind.data.frame(rep(4,times=length(cand4)), names(cand4), unname(cand4))

colnames(cand1) <- colnames(cand2) <- colnames(cand3) <- colnames(cand4) <- c("axis","snp","loading")

cand <- rbind(cand1, cand2, cand3, cand4)
cand$snp <- as.character(cand$snp)

#Add in the correlations of each candidate outlier SNP with the 5 environmental predictors

foo <- matrix(nrow=(ncand), ncol=5)  # 5 columns for 5 predictors
colnames(foo) <- c("long","depth","nitresursum","temperaturebot","salinitybot")

for (i in 1:length(cand$snp)) {
  nam <- cand[i,2]
  snp.gen <- geno_tur_tar_93.imp[,nam]
  foo[i,] <- apply(pred,2,function(x) cor(x,snp.gen))
}

cand <- cbind.data.frame(cand,foo)  
head(cand)

#Check if we have duplicate outlier detection on >1 axis

length(cand$snp[duplicated(cand$snp)])  # 0 duplicate detection

foo <- cbind(cand$axis, duplicated(cand$snp)) 
table(foo[foo[,1]==1,2]) # 0 duplicates on axis 1
table(foo[foo[,1]==2,2]) #  0 duplicates on axis 2
table(foo[foo[,1]==3,2]) #  0 duplicates on axis 3
table(foo[foo[,1]==4,2]) #  0 duplicates on axis 4

#Label the correlation between predictor and outlier SNP 

for (i in 1:length(cand$snp)) {
  bar <- cand[i,]
  cand[i,9] <- names(which.max(abs(bar[4:8]))) # gives the variable
  cand[i,10] <- max(abs(bar[4:8]))              # gives the correlation
}

colnames(cand)[9] <- "predictor"
colnames(cand)[10] <- "correlation"

table(cand$predictor) 

write.csv(cand, file="tur_tar_outlier_cand.csv")
```


RDA for P. turqueti samples from the Scotia Sea only:
```{R, error=TRUE, results='hide', eval=FALSE}
#RDA code from https://popgen.nescent.org/2018-03-27_RDA_GEA.html

library(psych)
library(vegan)
library(vcfR)
library(adegenet)
library(dartR)

#genotype file from -012 output of vcftools, using the thin1000_maf005_obshet05.vcf.recode.vcf file, from 01.SNP_filtering.Rmd
#only keep only Scotia Sea samples first with vcftools
geno_tur_tar_scotia <- read.csv("./tur_tar_scotia_5188_geno.csv", header=TRUE, row.names=1) 

str(geno_tur_tar_scotia)
dim(geno_tur_tar_scotia)

#RDA requires complete data frames (i.e., no missing data)

sum(is.na(geno_tur_tar_scotia)) #no NAs

#Read in the environmental factors
env <- read.csv("./scotia_tur_tar_env_var_GDM.csv")
str(env) # Look at the structure of the data frame
env$ID <- as.character(env$ID) # Make individual names characters (not factors)

#Confirm genotypes and environmental data are in the same order
identical(rownames(geno_tur_tar_scotia), env[,2])
as.list(rownames(geno_tur_tar_scotia))
as.list.data.frame(env[,2])
dim(env)

#check predictor correlations
png("./env_cor_panels.png", width=50, height=50, units="cm",res=155,pointsize = 26)
pairs.panels(env[,6:10], scale=T) #keep depth, longitude, bottom water salinity, bottom water temp_bottom, summer surface nitrate 
dev.off()

pred <- env[,6:10] #keep depth, longitude, bottom water salinity, bottom water temp_bottom, summer surface nitrate 

# Run the RDA

tur_scotia.rda <- rda(geno_tur_tar_scotia ~ ., data=pred, scale=T)
tur_scotia.rda

RsquareAdj(tur_scotia.rda)

summary(eigenvals(tur_scotia.rda, model = "constrained"))
screeplot(tur_scotia.rda) 

#Check RDA significance 
signif.full <- anova.cca(tur_scotia.rda, parallel=getOption("mc.cores")) # default is permutation=999
signif.full 

signif.axis <- anova.cca(tur_scotia.rda, by="axis", parallel=getOption("mc.cores"))
signif.axis

#Double check multicollinearity between predictors again 
vif.cca(tur_scotia.rda)

#Plot the RDA
plot(tur_scotia.rda, scaling=3)

#Plot RDA by locations
env$location <- factor(env$location,levels=c("ShagRocks","SouthGeorgia","SouthOrkney","ElephantIs","BransfieldStrait","KingGeorgeIs","RobertIs","LivingstonIs","DeceptionIs","Peninsula"))
spe <- env$location
bg <- c('#e6194B', '#3cb44b', '#ffe119', '#4363d8', '#f58231', '#42d4f4', '#f032e6', '#fabed4', '#469990', '#dcbeff', '#9A6324', '#fffac8', '#800000', '#aaffc3', '#000075', '#a9a9a9', '#ffffff', '#000000')

#axis 1 &2 
plot(tur_scotia.rda, type="n", scaling=3, family="sans", cex=0.75) 
points(tur_scotia.rda, display="species", pch=20, cex=0.7, col="gray32", scaling=3)  # the SNPs 
points(tur_scotia.rda, display="sites", pch=21, cex=1.3, col="gray32", scaling=3, bg=bg[spe])  # locations
text(tur_scotia.rda, scaling=3, display="bp", col="black", cex=0.7, family="sans")                           # the predictors
legend("bottomright", legend=levels(spe), xpd=TRUE, bty="n", col="gray32", pch=21, cex=0.75, pt.bg=bg)

#Identify candidate outlier SNPs

load.rda <- scores(tur.rda, choices=c(1:3), display="species")  # Species scores for the first constrained axes

#Axis 1-3 were selected as they were signficiant 

hist(load.rda[,1], main="Loadings on RDA1")
hist(load.rda[,2], main="Loadings on RDA2")
hist(load.rda[,3], main="Loadings on RDA3")

#define the function here as outliers

outliers <- function(x,z){
  lims <- mean(x) + c(-1, 1) * z * sd(x)     # find loadings +/-z sd from mean loading     
  x[x < lims[1] | x > lims[2]]               # locus names in these tails
}

#Apply outlier func to each significant constrained axis (3 SD)
cand1 <- outliers(load.rda[,1],3) # 15
cand2 <- outliers(load.rda[,2],3) # 35
cand3 <- outliers(load.rda[,3],3) # 30

ncand <- length(cand1) + length(cand2) + length(cand3) 
ncand #80 candidate outliers

cand1 <- cbind.data.frame(rep(1,times=length(cand1)), names(cand1), unname(cand1))
cand2 <- cbind.data.frame(rep(2,times=length(cand2)), names(cand2), unname(cand2))
cand3 <- cbind.data.frame(rep(3,times=length(cand3)), names(cand3), unname(cand3))

colnames(cand1) <- colnames(cand2) <- colnames(cand3) <- c("axis","snp","loading")

cand <- rbind(cand1, cand2, cand3)
cand$snp <- as.character(cand$snp)

#Add in the correlations of each candidate outlier SNP with the 5 environmental predictors

foo <- matrix(nrow=(ncand), ncol=5)  # 5 columns for 5 predictors
colnames(foo) <- c("long","depth","nitresursum","temperaturebot","salinitybot")

for (i in 1:length(cand$snp)) {
  nam <- cand[i,2]
  snp.gen <- geno_tur_tar_93.imp[,nam]
  foo[i,] <- apply(pred,2,function(x) cor(x,snp.gen))
}

cand <- cbind.data.frame(cand,foo)  
head(cand)

#Check if we have duplicate outlier detections on >1 axis

length(cand$snp[duplicated(cand$snp)])  # 0 duplicate detections

foo <- cbind(cand$axis, duplicated(cand$snp)) 
table(foo[foo[,1]==1,2]) # 0 duplicates on axis 1
table(foo[foo[,1]==2,2]) #  0 duplicates on axis 2
table(foo[foo[,1]==3,2]) #  0 duplicates on axis 3

#Label the correlation between predictor and outlier SNP 

for (i in 1:length(cand$snp)) {
  bar <- cand[i,]
  cand[i,9] <- names(which.max(abs(bar[4:8]))) # gives the variable
  cand[i,10] <- max(abs(bar[4:8]))              # gives the correlation
}

colnames(cand)[9] <- "predictor"
colnames(cand)[10] <- "correlation"

table(cand$predictor) 

write.csv(cand, file="tur_tar_scotia_outlier_cand")
```

