---
title: "06_outlier_loci_detection"
output: github_document
---

OutFLANK
```{R, error=TRUE, results='hide', eval=FALSE}
library(OutFLANK)
library(vcfR)
library(adegenet)
library(LEA)

#modify MakeDiploidFSTMat and getFSTs_diploids functions following here https://github.com/whitlock/OutFLANK/issues/19

##for all P. turqueti across the Southern Ocean 

tur_96_vcf <- read.vcfR("./thin1000_maf005_obshet05.vcf", verbose=FALSE)

#convert vcf to .geno format, .ped file got from vcftools --plink
tur_96_geno = ped2geno("./out.ped", output.file = "tur_96_geno", force = TRUE)
tur_96.geno_in <- read.fwf(tur_96_geno, width=rep(1,96))
tur_96.geno <- t(tur_96.geno_in)

tur_96_meta <- read.csv("tur_tar_popmap_96.csv", header=T) # Bring in the metadata with samples labelled by shelf vs scotia
tur_96_CLocus <- read.table("tur_96_5188_loci.txt", header=F) # Read CLocus ID (from --012 .pos file of vcftools)

## Calculate FST on the data
my_fst <- MakeDiploidFSTMat(tur_96.geno, locusNames = tur_96_CLocus$V1, popNames = tur_96_meta$region)
head(my_fst)

plot(my_fst$FST, my_fst$FSTNoCorr, 
     xlim=c(-0.01,0.3), ylim=c(-0.01,0.3),
     pch=20)

abline(0,1)

plot(my_fst$He, my_fst$FSTNoCorr, pch=20, col="grey")
hist(my_fst$FSTNoCorr, breaks=seq(0,0.6, by=0.001))
hist(my_fst$FSTNoCorr[my_fst$He>0.05])

#running outflank
out1 <- OutFLANK(FstDataFrame=my_fst, NumberOfSamples=2,
                 RightTrimFraction = 0.05, LeftTrimFraction = 0.05,
                 qthreshold = 0.01, Hmin = 0.1)

OutFLANKResultsPlotter(out1, withOutliers = TRUE,
                       NoCorr = TRUE, Hmin = 0.1, binwidth = 0.001, Zoom =
                         TRUE, RightZoomFraction = 0.2, titletext = NULL)

hist(out1$results$pvaluesRightTail)

outliers <- which(out1$results$OutlierFlag=="TRUE")

print(outliers)
vcfann <- as.data.frame(getFIX(tur_96_vcf))
outlierlist <- vcfann[outliers,]
write.csv(outlierlist, file="outlierlist_shelf_scotia.csv")


##for P. turqueti from the Scotia Sea

#keep samples from the Scotia Sea only via vcftools
tur_scotia_vcf <- read.vcfR("./thin1000_maf005_obshet05_scotia_only.vcf", verbose=FALSE)

#.ped file got from vcftools --plink
tur_scotia_geno = ped2geno("./out.ped", output.file = "tur_scotia_geno", force = TRUE)
tur_scotia.geno_in <- read.fwf(tur_scotia_geno, width=rep(1,52))
tur_scotia.geno <- t(tur_scotia.geno_in)

tur_scotia_meta <- read.table("tur_tar_scotiaonly_popmap.txt", header=F) # Bring in the metadata. Label samples by locations
tur_scotia_CLocus <- read.table("tur_96_5188_loci.txt", header=F) # Read CLocus ID (from --012 .pos file from vcftools)

## Calculate FST on the data
my_fst <- MakeDiploidFSTMat(tur_scotia.geno, locusNames = tur_scotia_CLocus$V1, popNames = tur_scotia_meta$V2)
head(my_fst)

plot(my_fst$FST, my_fst$FSTNoCorr, 
     xlim=c(-0.01,0.3), ylim=c(-0.01,0.3),
     pch=20)

abline(0,1)

plot(my_fst$He, my_fst$FSTNoCorr, pch=20, col="grey")

hist(my_fst$FSTNoCorr) #breaks=seq(0,0.9, by=0.001

hist(my_fst$FSTNoCorr[my_fst$He>0.05])

out1 <- OutFLANK(FstDataFrame=my_fst, NumberOfSamples=9,
                 RightTrimFraction = 0.05, LeftTrimFraction = 0.05,
                 qthreshold = 0.01, Hmin = 0.1)

OutFLANKResultsPlotter(out1, withOutliers = TRUE,
                       NoCorr = TRUE, Hmin = 0.1, binwidth = 0.001, Zoom =
                         TRUE, RightZoomFraction = 0.2, titletext = NULL)

hist(out1$results$pvaluesRightTail)

outliers <- which(out1$results$OutlierFlag=="TRUE")
print(outliers)
vcfann <- as.data.frame(getFIX(tur_scotia_vcf))
outlierlist <- vcfann[outliers,]
write.csv(outlierlist, file="outlierlist_scotia_location.csv")
```


BayeScan
```{R, error=TRUE, results='hide', eval=FALSE}
##for all P. turqueti across the Southern Ocean 
#label samples by shelf vs scotia
#input file generated by PGDSpider with thin1000_maf005_obshet05.vcf in 01.SNP_filtering.Rmd

singularity run $SING/bayescan-2.1.sif bayescan_2.1 turqueti_96_shelf_scotia_bayescan -n 100000 -thin 10 -nbp 20 -pilot 5000 -burn 50000 -pr_odds 100 -threads 10

#read Bayescan output
#code from http://evomicsorg.wpengine.netdna-cdn.com/wp-content/uploads/2016/01/BayeScan_BayeScEnv_exercises.pdf
#code from https://github.com/laurabenestan/Bayescan/wiki

library(coda)
chain<-read.table("./turqueti_96_shelf_scotia_bayescan.sel",header=TRUE)
chain<-chain[-c(1)]

#Create a MCMC object with the correct thinning interval 
chain<-mcmc(chain,thin=10)
plot(chain)
summary(chain)

#Check whether MCMC has converged
autocorr.diag(chain)
effectiveSize(chain)
geweke.diag(chain, frac1=0.1, frac2=0.5)
heidel.diag(chain, eps=0.1, pvalue=0.05)

#visualise Bayescan data for all P. turqueti across the Southern Ocean
library(ggplot2)
bayescan=read.table("./turqueti_96_shelf_scotia_bayescan_fst.txt")
SNPb_1=read.table("./out.012.pos",header=F)
SNPb=SNPb_1$V1

#Concatenate these two files
bayescan=cbind(SNPb, bayescan)

#Rename the columns
colnames(bayescan)=c("SNP","POST_PROB","LOG10_PO","Q_VALUE","ALPHA","FST")

#Use the information of the Q_value column
POST_PROB = 1 & Q_VALUE = 0 == 0.0001 
attach(bayescan)
class(bayescan$Q_VALUE)
bayescan$Q_VALUE <- as.numeric(bayescan$Q_VALUE)
bayescan[bayescan$Q_VALUE<=0.0001,"Q_VALUE"]=0.0001

#Specify the number of decimals
bayescan$POST_PROB <- (round(bayescan$POST_PROB, 4))
bayescan$LOG10_PO <- (round(bayescan$LOG10_PO, 4))
bayescan$Q_VALUE <- (round(bayescan$Q_VALUE, 4))
bayescan$ALPHA <- (round(bayescan$ALPHA, 4))
bayescan$FST <- (round(bayescan$FST, 6))

#Add a column SELECTION indicating the nature of each SNP : "diversifying", "neutral" or "balancing"
bayescan$SELECTION <- ifelse(bayescan$ALPHA>=0&bayescan$Q_VALUE<=0.01,"diversifying",ifelse(bayescan$ALPHA>=0&bayescan$Q_VALUE>0.01,"neutral","balancing"))
bayescan$SELECTION<- factor(bayescan$SELECTION)
levels(bayescan$SELECTION)

#Create subsets of each category of SNPs
positive <- bayescan[bayescan$SELECTION=="diversifying",]
neutral <- bayescan[bayescan$SELECTION=="neutral",]
balancing <- bayescan[bayescan$SELECTION=="balancing",]
xtabs(data=bayescan, ~SELECTION)

#Save these subsets into a .txt file

write.table(neutral, "./neutral_snps.txt", row.names=F, quote=F)
write.table(balancing, "./balancing_snps.txt", row.names=F, quote=F)
write.table(positive, "./positive_snps.txt", row.names=F, quote=F)



##for P. turqueti from the Scotia Sea only 
#label samples by locations
#input file generated by PGDSpider with thin1000_maf005_obshet05.vcf.recode.vcf in 01.SNP_filtering.Rmd

singularity run $SING/bayescan-2.1.sif bayescan_2.1 tur_tar_scotia_loc_bayescan -n 100000 -thin 10 -nbp 20 -pilot 5000 -burn 50000 -pr_odds 100 -threads 10

#Read the data
chain<-read.table("./tur_tar_scotia_loc_bayescan.sel",header=TRUE)
chain<-chain[-c(1)]

#Create a MCMC object with the correct thinning interval
chain<-mcmc(chain,thin=10)
plot(chain) 
summary(chain)

#Check whether MCMC has converged
autocorr.diag(chain)
effectiveSize(chain)
geweke.diag(chain, frac1=0.1, frac2=0.5)
heidel.diag(chain, eps=0.1, pvalue=0.05)

#visualise Bayescan data for P. turqueti from the Scotia Sea only 

library(ggplot2)
bayescan=read.table("./tur_tar_scotia_loc_bayescan_fst.txt")
SNPb_1=read.table("./out.012.pos",header=F)
SNPb=SNPb_1$V1

#Concatenate these two files
bayescan=cbind(SNPb, bayescan)

#Rename the columns
colnames(bayescan)=c("SNP","POST_PROB","LOG10_PO","Q_VALUE","ALPHA","FST")

#Use the information of the Q_value column
POST_PROB = 1 & Q_VALUE = 0 == 0.0001 
attach(bayescan)
class(bayescan$Q_VALUE)
bayescan$Q_VALUE <- as.numeric(bayescan$Q_VALUE)
bayescan[bayescan$Q_VALUE<=0.0001,"Q_VALUE"]=0.0001

#Specify the number of decimals
bayescan$POST_PROB <- (round(bayescan$POST_PROB, 4))
bayescan$LOG10_PO <- (round(bayescan$LOG10_PO, 4))
bayescan$Q_VALUE <- (round(bayescan$Q_VALUE, 4))
bayescan$ALPHA <- (round(bayescan$ALPHA, 4))
bayescan$FST <- (round(bayescan$FST, 6))

#Add a column SELECTION indicating the nature of each SNP : "diversifying", "neutral" or "balancing"
bayescan$SELECTION <- ifelse(bayescan$ALPHA>=0&bayescan$Q_VALUE<=0.01,"diversifying",ifelse(bayescan$ALPHA>=0&bayescan$Q_VALUE>0.01,"neutral","balancing"))
bayescan$SELECTION<- factor(bayescan$SELECTION)
levels(bayescan$SELECTION)

#Create subsets of each category of SNPs
positive <- bayescan[bayescan$SELECTION=="diversifying",]
neutral <- bayescan[bayescan$SELECTION=="neutral",]
balancing <- bayescan[bayescan$SELECTION=="balancing",]
xtabs(data=bayescan, ~SELECTION)

#Save these subsets into a .txt file

write.table(neutral, "./neutral_snps.txt", row.names=F, quote=F)
write.table(balancing, "./balancing_snps.txt", row.names=F, quote=F)
write.table(positive, "./positive_snps.txt", row.names=F, quote=F)
```

PCAdapt
```{R, error=TRUE, results='hide', eval=FALSE}
#code from https://bcm-uga.github.io/pcadapt/articles/pcadapt.html
#code from http://www.mountainmanmaier.com/software/pop_genom/

##for all P. turqueti across the Southern Ocean 
library(pcadapt)

vcftools --vcf thin1000_maf005_obshet05.vcf --plink
tur_96_pcadapt <- read.pcadapt("./out.ped", type = "ped")
x <- pcadapt(input = tur_96_pcadapt, K = 20)

tur_96_popmap <- read.csv("tur_tar_popmap_96.csv", header=T) # Bring in the popmap

#plot pcadapt 
plot(x, option = "screeplot") 

#Plot PCA on axis 1 and 2
plot(x,option="scores",i=1,j=2)

#Computing the test statistic based on PCA
x <- pcadapt(tur_96_pcadapt, K = 4)
summary(x)

#Manhattan Plot
plot(x , option = "manhattan")

#QQ Plot
plot(x, option="qqplot", threshold=0.05)

#Histogram of the test statistic and of the p-values
hist(x$pvalues, xlab = "p-values", main = NULL, breaks = 50, col = "orange")

#histogram of the test statistic 𝐷𝑗.
plot(x, option = "stat.distribution")

padj <- p.adjust(x$pvalues,method="BH")
alpha <- 0.01
outliers <- which(padj < alpha)
length(outliers) #766 outliers

loci <- read.table("./tur_96_5188_loci.txt", header=F)
outliers1 <- as.character(loci[,1])[outliers]
outliers1 #766 outliers

write.csv(outliers1, file="tur_all_outliers.csv")


##for P. turqueti from the Scotia Sea  

singularity run $SING/vcftools-0.1.16.sif vcftools --vcf thin1000_maf005_obshet05_scotiaonly.vcf --plink #only keep samples from the Scotia Sea

tur_scotia_pcadapt <- read.pcadapt("./out.ped", type = "ped")

x <- pcadapt(input = tur_scotia_pcadapt, K = 20) #this line assumes indv is diploid

tur_scotia_popmap <- read.table("tur_tar_scotiaonly_popmap.txt", header=F) # Bring in the popmap

#plot pcadapt 
plot(x, option = "screeplot")

#Plot PCA on axis 1 and 2
plot(x,option="scores",i=1,j=2)

#Computing the test statistic based on PCA
x <- pcadapt(tur_scotia_pcadapt, K = 5)
summary(x)

#Manhattan Plot
plot(x , option = "manhattan")

#QQ Plot
plot(x, option="qqplot", threshold=0.01)

#Histograms of the test statistic and of the p-values
hist(x$pvalues, xlab = "p-values", main = NULL, breaks = 50, col = "orange")

#histogram of the test statistic 𝐷𝑗.
plot(x, option = "stat.distribution")

padj <- p.adjust(x$pvalues,method="BH")
alpha <- 0.01
outliers <- which(padj < alpha)
length(outliers) #429 outliers

loci <- read.table("./tur_96_5188_loci.txt", header=F) #same loci list as before
outliers1 <- as.character(loci[,1])[outliers]
outliers1 #429 outliers

write.csv(outliers1, file="tur_scotia_loc_outliers.csv")
```
